### PR TITLE
Cell sets

### PR DESCRIPTION
**Pull request by keller-mark**
This adds processing for cell sets files according to the schema defined in https://github.com/hubmapconsortium/vitessce/pull/558

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -9,4 +9,5 @@ pyimzml==1.3.0
 dask==2.11.0
 toolz==0.9.0
 fsspec==0.3.3
-tifffile==2020.2.16
\ No newline at end of file
+tifffile==2020.2.16
+scikit-image==0.16.1
```
Comment (ilan-gold) [inline requirements.txt:13]: Why this?

File: fake-files/output-expected/linnarsson/linnarsson.cell_sets.json [hunk 2]
```diff
@@ -0,0 +1 @@
..., "2782", "3143", "1613", "4746", "5657", "334", "1843", "614", "5583", "229", "3797", "1800", "4530", "4415", "972", "2365", "3625", "4405", "337", "54", "1742", "3423", "3172", "1698", "7825", "1716", "3186", "6586", "5152", "2966", "3159", "290", "6541", "2316"]}, {"name": "Ependymal", "set": ["3184", "3146", "2707", "1201", "3107", "1087", "3129", "1181", "1191", "5366", "3126", "3182", "3151", "4411", "1182", "1179", "3198", "3218", "2604", "1383", "3115", "2788", "1189", "2704", "2896", "2901", "3140", "2663", "5371", "3122", "3138", "3210", "2713", "1202", "4277", "3176", "4185", "1194", "5455", "3167", "1192", "3104", "3201", "3188", "3171", "2902", "2928", "3127", "3202", "3205", "5630", "3112", "3137", "3183", "3211", "1498", "3217", "1200", "1186", "3135", "3161", "3175", "7067", "2924", "3118", "1184", "3130", "3134", "3121", "2722", "2899", "2654", "2720", "3105", "3136", "4262", "3133", "1190", "1185", "4266", "2652", "6595", "3124", "3204", "3132", "1196", "3148", "2716", "1195", "3152", "3199", "3200", "3123", "3147", "2965", "3187", "3120", "3180", "3131", "3220", "1090", "3179", "3209", "3178", "5398", "5408", "2648", "4256", "3197", "1382", "5397", "1381"]}]}]}]}
```
Comment (ilan-gold) [inline fake-files/output-expected/linnarsson/linnarsson.cell_sets.json]: Why is this here now?  I don't mind that it's not really a "unit" test but I'm just curious.

**Review by ilan-gold — APPROVED** [review]
I'm OK with this whole process, but I would like to know the fate of factors one way or the other before getting this into Vitessce.

File: requirements.txt (hunk 1 above)
Comment (keller-mark) [reply requirements.txt:13]: The builds were failing on Travis https://travis-ci.com/github/hubmapconsortium/vitessce-data/builds/166386020#L288

First I tried adding 0.17.1 (the latest scikit-image) https://github.com/hubmapconsortium/vitessce-data/pull/94/commits/22f3eac8ad8a386301c015b8c95d8660d00c9556 but that failed too.

When I downgraded to 0.16.1 https://github.com/hubmapconsortium/vitessce-data/pull/94/commits/a63f92c9ebbed7036d6ef5a5f744d8c76dd8dab8 it worked.

File: fake-files/output-expected/linnarsson/linnarsson.cell_sets.json (hunk 2 above)
Comment (keller-mark) [reply fake-files/output-expected/linnarsson/linnarsson.cell_sets.json]: This follows the current convention of the Linnarsson files in the fake-files folder actually being real files. Not sure why either https://github.com/hubmapconsortium/vitessce-data/tree/master/fake-files/output-expected/linnarsson

File: python/cell_reader.py [hunk 3]
```diff
@@ -195,6 +196,69 @@ def get_factors(metadata):
     return factors
 
 
+def get_cell_sets(clusters, lookup):
+    '''
```
Comment (manzt) [inline python/cell_reader.py:199]: not sure if the `namedtuple` import can be done within the doctest, but does this work?

```suggestion
    '''
    >>> from collections import namedtuple
```

File: python/cell_reader.py [hunk 4]
```diff
@@ -195,6 +196,69 @@ def get_factors(metadata):
...
+        (c.name, c.cell_ids) for c in clusters.values()
+    )
+
+    hierarchy = dict([
+        (c, dict([
+            (sc, cluster_name_to_cell_ids[sc])
+            for sc, sc_c in lookup.items() if sc_c == c
+        ]))
+        for c in lookup.values()
+    ])
```
Comment (manzt) [inline python/cell_reader.py]: Took be a bit to understand what was going on here. Would something like this work? 

Probably not that important, but having shared variable names between the creation and traversal of the hierarchy might helpful for future readers, and lookup is used as a the hash table. If failing loudly is desired, I'd replace `Dictionary.get` with `Dictionary[]`, since the former returns `None` if a key is missing.

```suggestion
    hierarchy = {cluster_name: {} for cluster_name in lookup.values()}
    for c in clusters.values():
        subcluster = {c.name: c.cell_ids}
        cluster_name = lookup.get(c.name)
        cluster_dict = hierarchy.get(cluster_name)
        cluster_dict.update(subcluster)
```

**Review by manzt — APPROVED** [review]
Looks good. Just some comments on the hierarchy construction, take or leave!

File: python/cell_reader.py (hunk 4 above)
Comment (keller-mark) [reply python/cell_reader.py]: Thanks for the suggestion! I have adopted this change and added comments to describe more clearly what the code is doing / what the variables contain in the latest commit https://github.com/hubmapconsortium/vitessce-data/pull/94/commits/28ea851898e1d58c5cd3725709de77f9b29cc915

File: fake-files/output-expected/linnarsson/linnarsson.cell_sets.json (hunk 2 above)
Comment (mccalluc) [reply fake-files/output-expected/linnarsson/linnarsson.cell_sets.json]: I'm not sure I really get the question or the answer... ping me if I should?

Comment (mccalluc) [reply fake-files/output-expected/linnarsson/linnarsson.cell_sets.json]: but:
- please pretty print this.
- Consider making it smaller: start with a smaller test input.... if there is a difference in the output, you want the diff to be readable.

File: python/cell_reader.py [hunk 5]
```diff
@@ -195,6 +195,80 @@ def get_factors(metadata):
...
+    # ```
+    # {
+    #   Cluster A: {
+    #     Subcluster A: [1, 2],
+    #     Subcluster B: [3, 4]
+    #   },
+    #   Cluster B: {...}
+    # }
+    # ```
+    hierarchy = {cluster_name: {} for cluster_name in lookup.values()}
```
Comment (mccalluc) [inline python/cell_reader.py:236]: Might use a defaultdict here, but this works, too.

File: python/cell_reader.py [hunk 6]
```diff
@@ -195,6 +195,80 @@ def get_factors(metadata):
...
+    for cluster_name in sorted(hierarchy.keys()):
+        cluster_dict = hierarchy[cluster_name]
+        subcluster_nodes = []
+        for subcluster_name in sorted(cluster_dict.keys()):
+            subcluster = cluster_dict[subcluster_name]
+            subcluster_nodes.append({
+                'name': subcluster_name,
+                'set': subcluster
+            })
+        cluster_nodes.append({
```
Comment (mccalluc) [inline python/cell_reader.py:255]: consider a helper function and array comprehensions when things get a little deep.

File: python/cell_reader.py [hunk 7]
```diff
@@ -311,6 +389,12 @@ def add_pca(metadata):
     if args.cells_file:
         json.dump(metadata, args.cells_file, indent=1)
 
+    if args.cell_sets_file:
+        clusters = lr.clusters()
+        cell_sets = get_cell_sets(clusters, LOOKUP)
+        cell_sets_json = json.dumps(cell_sets)
```
Comment (mccalluc) [inline python/cell_reader.py]: pretty print?

File: python/dries_json_reader.py [hunk 8]
```diff
@@ -48,6 +48,75 @@ def cells_json(data):
...
+    ... }
+    >>> cell_sets = cell_sets_json(data)
+    >>> list(cell_sets.keys())
+    ['version', 'datatype', 'tree']
+    >>> cell_sets['datatype']
+    'cell'
+    >>> len(cell_sets['tree'])
+    2
+    >>> sorted([ n['name'] for n in cell_sets['tree'] ])
+    ['Leiden Clustering', 'k-means Clustering']
```
Comment (mccalluc) [inline python/dries_json_reader.py:75]: (If it ever feels like this is more that should be crammed in a doctest, feel free to set up a standalone.: I do like to keep tests close, but if you find yourself doing actual coding inside the doctest, it can get painful.)

File: python/dries_json_reader.py [hunk 9]
```diff
@@ -48,6 +48,75 @@ def cells_json(data):
...
+    '''
+    clustering_dict = {
+        'pleiden_clus': {},
+        'kmeans': {}
+    }
+    nice_names = {
+        'pleiden_clus': 'Leiden Clustering',
```
Comment (mccalluc) [inline python/dries_json_reader.py:82]: This is just for our demos... but I would steer clear of making things prettier than what the data can provide. If the demos are ugly, that's some pressure on the labs to give us better data.

File: python/dries_json_reader.py [hunk 10]
```diff
@@ -48,6 +48,75 @@ def cells_json(data):
...
+        'kmeans': 'k-means Clustering'
+    }
+    for cell_id in data.keys():
+        factors = data[cell_id]['factors']
+        factors_dict = {
+            'pleiden_clus': 'Cluster {}'.format(factors['pleiden_clus'][0]),
+            'kmeans': 'Cluster {}'.format(factors['kmeans'][0])
```
Comment (mccalluc) [inline python/dries_json_reader.py:89]: The parallelism makes me think something could be factored out?

File: python/dries_json_reader.py [hunk 11]
```diff
@@ -48,6 +48,75 @@ def cells_json(data):
...
+        }
+        # For each cluster assignment, append this cell ID to the
+        # appropriate clustering_dict list.
+        for factor_type, factor_cluster in factors_dict.items():
+            if factor_cluster in clustering_dict[factor_type]:
+                clustering_dict[factor_type][factor_cluster].append(cell_id)
+            else:
+                clustering_dict[factor_type][factor_cluster] = [cell_id]
```
Comment (mccalluc) [inline python/dries_json_reader.py]: This is something a defaultdict could clean up.

File: python/dries_json_reader.py [hunk 12]
```diff
@@ -48,6 +48,75 @@ def cells_json(data):
...
+    # Construct the tree
+    cell_sets = {
+        'version': '0.1.2',
+        'datatype': 'cell',
+        'tree': []
+    }
+
+    for factor_type, factor_clusters in clustering_dict.items():
+        factor_type_children = []
+        for cluster_name in sorted(factor_clusters.keys()):
```
Comment (mccalluc) [inline python/dries_json_reader.py:106]: Use comprehension instead?

**Review by mccalluc — COMMENTED** [review]
More little comments. Thanks!

File: python/dries_json_reader.py (hunk 8 above)
Comment (keller-mark) [reply python/dries_json_reader.py:75]: Sounds good!

File: python/dries_json_reader.py (hunk 9 above)
Comment (keller-mark) [reply python/dries_json_reader.py:82]: Right, I think it is OK for the demo data here.

File: python/cell_reader.py (hunk 6 above)
Comment (keller-mark) [reply python/cell_reader.py:255]: I got previous feedback that the array comprehension was more difficult to understand. Ultimately I think the tests are helpful here to ensure that the code does what it is supposed to do, regardless of implementation https://github.com/hubmapconsortium/vitessce-data/pull/94#discussion_r426164341

File: python/dries_json_reader.py (hunk 12 above)
Comment (keller-mark) [reply python/dries_json_reader.py:106]: Please see previous comment about original comprehension implementation being more difficult to understand.

File: fake-files/output-expected/linnarsson/linnarsson.cell_sets.json (hunk 2 above)
Comment (keller-mark) [reply fake-files/output-expected/linnarsson/linnarsson.cell_sets.json]: I have added pretty printing here.

To clarify my previous answer:
This is a new test for the `linnarsson` dataset.

The existing tests for all of the Linnarsson data in this vitessce-data repository use the FULL linnarsson.cells.loom file. https://github.com/hubmapconsortium/vitessce-data/tree/master/fake-files/input/linnarsson

To change the size of this test fixture would mean changing all of the other test fixtures for all of the other linnarsson tests. Do you want me to do that here?
